File: gmstudio/gmsampler/noise.go
package gmsampler

import "github.com/MaxKlaxxMiner/golangMusicStudio/gmstudio/gmconst"

func Noise(buf []int32, incr, ofs uint32) uint32 {
	for i := 0; i < len(buf); i++ {
		buf[i] = int32(ofs<<(gmconst.SampleBits-gmconst.DynamicBits-4)) >> (gmconst.SampleBits - gmconst.DynamicBits)
		ofs = (ofs+incr)*214013 + 2531011
	}
	return ofs
}

File: gmstudio/gmsampler/noise.c
// wasm-compiler: https://wasdk.github.io/WasmFiddle/
typedef int int32;
typedef unsigned int uint32;
typedef long long int64;
typedef unsigned long long uint64;
#define SampleBits 32
#define DynamicBits 24

uint32 noise(int32* buf, int sampleCount, uint32 incr, uint32 ofs) {
  for (int i = 0; i < sampleCount; i++) {
    buf[i] = (int32)(ofs << (SampleBits - DynamicBits - 4)) >> (SampleBits - DynamicBits);
    ofs = (ofs + incr) * 214013 + 2531011;
  }
  return ofs;
}
